connector/keystone/keystone.go

```diff
@@ -0,0 +1,189 @@
+// Package keystone provides authentication strategy using Keystone.
+package keystone
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/dexidp/dex/connector"
+	"github.com/sirupsen/logrus"
+	"io/ioutil"
+	"net/http"
+)
+
+var (
+	_ connector.PasswordConnector = &keystoneConnector{}
+	_ connector.RefreshConnector  = &keystoneConnector{}
+)
+
+// Open returns an authentication strategy using Keystone.
+func (c *Config) Open(id string, logger logrus.FieldLogger) (connector.Connector, error) {
+	return &keystoneConnector{c.Domain, c.KeystoneHost,
+		c.KeystoneUsername, c.KeystonePassword, logger}, nil
+}
+
+func (p keystoneConnector) Close() error { return nil }
+
+func (p keystoneConnector) Login(ctx context.Context, s connector.Scopes, username, password string) (
+	identity connector.Identity, validPassword bool, err error) {
+	resp, err := p.getTokenResponse(ctx, username, password)
+
+	if err != nil {
+		return identity, false, fmt.Errorf("keystone: error %v", err)
+	}
+
+	// Providing wrong password or wrong keystone URI throws error
+	if resp.StatusCode == 201 {
+		token := resp.Header.Get("X-Subject-Token")
+		data, err := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+
+		if err != nil {
+			return identity, false, err
+		}
+
+		var tokenResponse = new(tokenResponse)
```

Comment:
[nit] I'm not a fan of clashing names between types and variables; but if you are, that's tine 😉